komga/src/main/kotlin/org/gotson/komga/domain/service/LibraryScanner.kt

```diff
@@ -18,95 +26,238 @@ class LibraryScanner(
+        if (existingSeries.deleted) {
+          seriesLifecycle.deleteOne(existingSeries.id)
           logger.info { "Adding new series: $newSeries" }
-          val createdSeries = seriesLifecycle.createSeries(newSeries)
-          seriesLifecycle.addBooks(createdSeries, newBooks)
-          seriesLifecycle.sortBooks(createdSeries)
-        } else {
-          // if series already exists, update it
-          logger.debug { "Scanned series already exists. Scanned: $newSeries, Existing: $existingSeries" }
-          val seriesChanged = newSeries.fileLastModified.truncatedTo(ChronoUnit.MILLIS) != existingSeries.fileLastModified.truncatedTo(ChronoUnit.MILLIS)
-          if (seriesChanged) {
-            logger.info { "Series changed on disk, updating: $existingSeries" }
-            seriesRepository.update(existingSeries.copy(fileLastModified = newSeries.fileLastModified))
+          seriesLifecycle.createSeries(newSeries)
+          return@let
+        }
+
+        logger.debug { "Scanned series already exists. Scanned: $newSeries, Existing: $existingSeries" }
+        if (isModified(newSeries.fileLastModified, existingSeries.fileLastModified)) {
+          logger.info { "Series changed on disk, updating: $existingSeries" }
+          seriesRepository.update(existingSeries.copy(fileLastModified = newSeries.fileLastModified))
+        }
+
+        if (library.scanDeep || isModified(newSeries.fileLastModified, existingSeries.fileLastModified)) {
+          val existingBooks = bookRepository.findBySeriesId(existingSeries.id)
+          logger.debug { "Existing books: $existingBooks" }
+          updateExistingBook(newBooks, existingBooks)
+          deleteNoLongerExistingBooks(newBooks, existingBooks)
+        }
+      } ?: run {
+        logger.info { "Adding new series: $newSeries" }
+        seriesLifecycle.createSeries(newSeries)
+      }
+    }
+  }
+
+  private fun restoreDeletedBooksAndAddNew(scannedSeries: Map<Series, Collection<Book>>, library: Library) {
+    scannedSeries.forEach { (newSeries, newBooks) ->
+      seriesRepository.findByLibraryIdAndUrlIncludeDeleted(library.id, newSeries.url)?.let { existingSeries ->
+        val existingBooks = bookRepository.findBySeriesId(existingSeries.id)
+        val restoredBooks = restoreDeletedBooks(existingSeries, existingBooks, newBooks)
+
+        val booksToAdd = newBooks
+          .filterNot { newBook -> existingBooks.map { it.url }.contains(newBook.url) }
+          .filterNot { toCheck -> restoredBooks.map { it.fileHash }.contains(toCheck.fileHash) }
+
+        if (booksToAdd.isNotEmpty()) {
+          logger.info { "Adding new books: $booksToAdd" }
+          seriesLifecycle.addBooks(existingSeries, booksToAdd)
+          seriesLifecycle.sortBooks(existingSeries)
+        }
+      }
+    }
+  }
+
+  private fun restoreDeletedSeries(scannedSeries: Map<Series, Collection<Book>>) {
```

Comment:
Can you explain what's the strategy around restoring series ?

I would have thought that restoring books should be enough, and then we could matche the deleted series by the book's `seriesId`.

You seem to be doing something more complex though.